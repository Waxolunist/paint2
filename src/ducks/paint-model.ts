export interface Painting {
  id: number;
  dataUrl: string;
  blobUrl?: string;
}

export interface Point {
  x: number;
  y: number;
}

export interface Stroke {
  points: Point[];
  erase: boolean;
  color: string;
}

export interface PaintingRawData {
  paintingId: number;
  strokes: Stroke[];
}

export interface CRUDPayload {
  painting: Painting;
  rawData: PaintingRawData;
}

export interface PaintState {
  paintings: Painting[];
  activePainting?: CRUDPayload;
}
