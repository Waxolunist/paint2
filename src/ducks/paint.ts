import {
  PaintState,
  CRUDPayload,
  PaintingImpl,
  PaintingRawData,
  Painting,
} from './paint-model';
import {AnyAction, Reducer} from 'redux';
import {ThunkDispatch as TDispatch, ThunkAction as TAction} from 'redux-thunk';
import database from '../database';

/*** types ***/
const STORE = '@paint/STORE';
const LOAD = '@paint/LOAD';
const NEW = '@paint/NEW';
const DELETE = '@paint/DELETE';

/*** actions ***/
interface CRUDAction extends AnyAction {
  type: typeof STORE | typeof LOAD | typeof NEW | typeof DELETE;
  payload: CRUDPayload | number | string | undefined;
}

type PaintActionTypes = CRUDAction;

export type ThunkDispatch = TDispatch<PaintState, null, PaintActionTypes>;

export type ThunkAction = TAction<void, PaintState, null, PaintActionTypes>;

export const storeData = ({
  id,
  dataUrl,
  strokes,
}: {
  id?: string;
  dataUrl: string;
  strokes: Stroke[];
}): ThunkAction => async (dispatch) => {
  const db = await database();
  const painting = new PaintingImpl(dataUrl, id);
  const paintingId = await db.paintings.put(painting);
  await painting.cleanState();
  const rawData = new PaintingRawData(paintingId, strokes);
  await db.strokes.put(rawData);
  return dispatch({
    type: STORE,
    payload: {
      painting,
      rawData,
    },
  });
};

export const loadData = (id?: number | string): ThunkAction => async (dispatch) => {
  if(id) {
    const db = await database();
    const painting = await db.paintings.get(parseInt(id as string));
    if (painting) {
      return dispatch({
        type: LOAD, payload: {
          painting,
          rawData: undefined,
        }
      });
    }
  }
  return undefined;
};

export const newPainting = (): ThunkAction => async (dispatch) => {
  const db = await database();
  const painting = new PaintingImpl();
  await db.paintings.put(painting);
  return dispatch({
    type: NEW,
    payload: {
      painting,
    },
  });
};

export const removePainting = (id?: number | string): ThunkAction => async dispatch => {
  if(id) {
    const db = await database();
    await db.paintings.delete(parseInt(id as string))
    return dispatch({
      type: DELETE,
      payload: id
    });
  }
  return undefined;
};

const initialState: PaintState = {
  paintings: [],
  activePainting: undefined,
};

const paintingsArray = (paintings: Painting[]): Painting[] => {
  return [
    ...new Map(
      paintings
        .filter((p) => p)
        .map((p: Painting): [number | undefined, Painting] => [p.id, p])
        .sort(([idA], [idB]) => idA! - idB!)
    ).values(),
  ];
};

const removePaintingFromArray = (id: number | string, paintings: Painting[]): Painting[] => paintings.filter(p => p.id != id);

const paintReducer: Reducer<PaintState, AnyAction> = (
  state = initialState,
  action: AnyAction
): PaintState => {
  switch (action.type) {
    case STORE:
      return {
        ...state,
        paintings: paintingsArray([
          ...state.paintings,
          ((<CRUDAction>action).payload as CRUDPayload).painting,
        ]),
        activePainting: undefined,
      };
    case LOAD:
    case NEW:
      return {
        ...state,
        paintings: paintingsArray([
          ...state.paintings,
          ((<CRUDAction>action).payload as CRUDPayload).painting,
        ]),
        activePainting: (<CRUDAction>action).payload as CRUDPayload,
      };
    case DELETE:
      return {
        ...state,
        paintings: removePaintingFromArray((<CRUDAction>action).payload as string, state.paintings),
        activePainting: undefined,
      };
    default:
      return state;
  }
};

export default paintReducer;
