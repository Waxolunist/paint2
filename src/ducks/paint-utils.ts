import {Painting} from './paint-model';

export const extractIdFromUrl = (): string | undefined => {
  const match = window.location.pathname.match(/\/paint\/(\d+)/);
  return match ? match[1] : undefined;
};

export const removePaintingFromArray = (
  id: number | string,
  paintings: Painting[]
): Painting[] => paintings.filter((p) => p.id != id);

export const newSortedDeduplicatedPaintingsArray = (
  paintings: Painting[]
): Painting[] => [
  ...new Map(
    paintings
      .filter((p: Painting | undefined) => p)
      .map((p: Painting): [number | undefined, Painting] => [p.id, p])
      .sort(
        (
          [idA]: [number | undefined, Painting],
          [idB]: [number | undefined, Painting]
        ) => (idA ?? 0) - (idB ?? 0)
      )
  ).values(),
];
