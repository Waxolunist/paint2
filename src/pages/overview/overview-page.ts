import {LitElement, html, css, CSSResult, TemplateResult} from 'lit';
import {customElement, property, queryAll} from 'lit/decorators.js';
import {styleMap} from 'lit/directives/style-map.js';
import '../../components/new-paint-button/new-paint-button';
import '../../components/paint-button/paint-button';
import '../../components/icon-button/paint-icon-button';
import {closeIcon, shareIcon} from './icons';
import store from '../../store';
import {connect} from 'pwa-helpers/connect-mixin';
import {Painting, PaintState, defaultDataUrl} from '../../ducks/paint-model';
import {repeat} from 'lit/directives/repeat.js';
import {ifDefined} from 'lit/directives/if-defined.js';
import {
  loadData,
  newPainting,
  ThunkDispatch,
  removePainting,
} from '../../ducks/paint';
import {toFileExtension, dataURLtoBlob} from '../../ducks/paint-utils';
import {RouterState} from 'lit-redux-router/lib/reducer';
import {AnimatedStyles} from '../../styles/shared-styles';

interface PaintingAnimationData {
  id: number;
  index: number;
  clientRect: DOMRect;
}

@customElement('paint-overview-page')
export class OverviewPage extends connect(store)(LitElement) {
  @queryAll('paint-paint-button')
  paintingButtons?: NodeList;

  private currentPaintingClientRects: PaintingAnimationData[] = [];

  @property({type: Array, attribute: false})
  paintings: Painting[] = [];

  stateChanged({
    router,
    paint,
  }: {
    router: RouterState;
    paint: PaintState;
  }): void {
    if (router.routes['/']?.active) this.paintings = paint.paintings;
  }

  updated(changedProperties: Map<string, Painting[]>): void {
    if (this.paintingDeleted(changedProperties.get('paintings'))) {
      this.updateComplete.then(this.animatePaintings);
    }
  }

  static styles: CSSResult[] = [
    //language=CSS
    AnimatedStyles,
    css`
      :host {
        display: block;
        background-color: #91b5ff;
        padding: 1.5em 0;
        /* prettier-ignore */
        min-width: calc((var(--painting-width) / var(--painting-scalefactor) + var(--painting-margin)) * 2);
        position: relative;
      }

      .paintings {
        display: grid;
        grid-gap: var(--painting-margin);
        grid-template-columns: repeat(
          auto-fit,
          calc(var(--painting-width) / var(--painting-scalefactor))
        );
        justify-content: center;
      }

      .painting {
        width: calc(var(--painting-width) / var(--painting-scalefactor));
        height: calc(var(--painting-height) / var(--painting-scalefactor));
        position: relative;
        transition-property: transform;
        will-change: transform;
      }

      .painting.translate {
        transform: translate(0px, 0px);
      }

      .about {
        position: absolute;
        bottom: -2em;
        padding: 1em;
        font-size: large;
        font-family: monospace;
        font-weight: bold;
      }

      .about a {
        color: #ffffef;
      }

      @media screen and (max-width: 400px) {
        :host {
          padding: 1.5em 0;
          --painting-margin: 1em;
        }

        .paintings {
          --painting-scalefactor: 1.3;
          grid-column-gap: 1.3em;
          grid-row-gap: 1.3em;
        }
      }
    `,
  ];

  render(): TemplateResult {
    return html`
      <div class="paintings">
        <paint-new-paint-button
          class="painting translate"
          @paint-clicked="${this.newPainting}"
        ></paint-new-paint-button>
        ${repeat(
          this.paintings,
          (painting, index) =>
            html`<paint-paint-button
              class="painting translate"
              data-id="${ifDefined(painting.id)}"
              style="${styleMap(
                this.calculateTranslateProperty(index, painting.id)
              )}"
              @paint-clicked="${this.openPainting(painting.id)}"
              imageUrl="${painting.blobUrl}"
            >
              <paint-icon-button
                role="button"
                slot="addons"
                class="delete"
                @icon-clicked="${this.removePainting(painting.id)}"
                >${closeIcon}</paint-icon-button
              >
              ${this.canShare()
                ? html`<paint-icon-button
                    slot="addons"
                    role="button"
                    class="share"
                    @icon-clicked="${this.sharePainting(painting.id)}"
                    >${shareIcon}</paint-icon-button
                  >`
                : ''}
            </paint-paint-button>`
        )}
      </div>
      <div class="about"><a href="/about">About this app</a></div>
    `;
  }

  private newPainting = (): void => {
    (<ThunkDispatch>store.dispatch)(newPainting());
  };

  private openPainting = (id?: number | string) => (): void => {
    (<ThunkDispatch>store.dispatch)(loadData(id));
  };

  private removePainting =
    (id?: number | string) => async (): Promise<void> => {
      this.currentPaintingClientRects = [];
      this.paintingButtons?.forEach((node, index) => {
        this.currentPaintingClientRects.push({
          id: this.paintings[index].id ?? 0,
          index: index,
          clientRect: (node as HTMLElement).getBoundingClientRect(),
        });
      });
      this.removeAnimatedClass();
      (<ThunkDispatch>store.dispatch)(removePainting(id));
    };

  private canShare = (): boolean | undefined => {
    const blob = dataURLtoBlob(defaultDataUrl);
    return (
      navigator.canShare &&
      navigator.canShare({
        files: [
          new File([blob], `test.${toFileExtension(blob.type)}`, {
            type: blob.type,
          }),
        ],
      })
    );
  };

  private sharePainting = (id?: number | string) => async (): Promise<void> => {
    const painting: Painting | undefined = store
      .getState()
      .paint.paintings.find((p) => p.id == id);
    if (painting) {
      const request = await fetch(painting.blobUrl);
      const blob = await request.blob();
      try {
        await navigator.share({
          files: [
            new File([blob], `painting.${toFileExtension(blob.type)}`, {
              type: blob.type,
            }),
          ],
          title: 'My Painting',
          text: `Look Ma' what I did.`,
          url: '/',
        });
      } catch (error) {
        console.error('Sharing failed', error);
      }
    }
  };

  private paintingDeleted = (oldVal: Painting[] = []): boolean =>
    !!oldVal && oldVal.length > this.paintings.length;

  private calculateTranslateProperty = (
    index: number,
    id?: number
  ): {[name: string]: string} => {
    const dataWithIndexAndId = this.currentPaintingClientRects.find(
      (p) => p.index === index && p.id === id
    );
    if (dataWithIndexAndId || this.currentPaintingClientRects.length === 0) {
      return {transform: 'translate(0px, 0px)'};
    }
    const dataWithIdIndex = this.currentPaintingClientRects.findIndex(
      (p) => p.id === id
    );
    const current = this.currentPaintingClientRects[dataWithIdIndex];
    const predeccessor = this.currentPaintingClientRects[dataWithIdIndex - 1];
    if (current && predeccessor) {
      const translateX = current.clientRect.x - predeccessor.clientRect.x;
      const translateY = current.clientRect.y - predeccessor.clientRect.y;
      return {transform: `translate(${translateX}px, ${translateY}px)`};
    }
    return {transform: 'translate(0px, 0px)'};
  };

  private animatePaintings = (): void => {
    this.paintingButtons?.forEach((node, index) => {
      const htmlel = node as HTMLElement;
      setTimeout(
        () =>
          requestAnimationFrame(() => {
            htmlel.style.transform = 'translate(0px, 0px)';
            htmlel.classList.add('animated');
          }),
        Math.min(200 + index * 50, 700)
      );
    });
  };

  private removeAnimatedClass = (): void => {
    this.paintingButtons?.forEach((node) =>
      (node as HTMLElement).classList.remove('animated')
    );
  };
}
