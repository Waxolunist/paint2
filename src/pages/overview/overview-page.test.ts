import {elementUpdated, fixture, html} from '@open-wc/testing';
import {cleanHTML} from '../../test/htmlutils';
import {OverviewPage} from './overview-page';
import {PaintingImpl} from '../../ducks/paint-model';

describe('overview-page', () => {
  it('is defined', () => {
    const el = document.createElement('paint-overview-page');
    expect(el).toBeInstanceOf(OverviewPage);
  });

  it('renders correctly', async () => {
    const element = await fixture(
      html` <paint-overview-page></paint-overview-page>`
    );
    expect(cleanHTML(element)).toMatchSnapshot();
  });

  it('renders correctly with paintings', async () => {
    const element: OverviewPage = await fixture(
      html` <paint-overview-page></paint-overview-page>`
    );
    element.paintings = [new PaintingImpl(), new PaintingImpl()];
    await elementUpdated(element);
    expect(cleanHTML(element)).toMatchSnapshot();
    expect(
      element.shadowRoot?.querySelectorAll('paint-paint-button')
    ).toHaveLength(2);
  });
});
