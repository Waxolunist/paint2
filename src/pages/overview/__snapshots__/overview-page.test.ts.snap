// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`overview-page renders correctly 1`] = `
"
      <div class="paintings">
        <paint-new-paint-button class="painting translate"></paint-new-paint-button>
        
      </div>
      <div class="about"><a href="/about">About this app</a></div>
    <style>
  .animated {
    transition-duration: 0.28s;
    transition-timing-function: cubic-bezier(0.4, 0, 0.2, 1);
  }
</style><style>
      :host {
        display: block;
        background-color: #91b5ff;
        padding: 1.5em 0;
        /* prettier-ignore */
        min-width: calc((var(--painting-width) / var(--painting-scalefactor) + var(--painting-margin)) * 2);
        position: relative;
      }

      .paintings {
        display: grid;
        grid-gap: var(--painting-margin);
        grid-template-columns: repeat(
          auto-fit,
          calc(var(--painting-width) / var(--painting-scalefactor))
        );
        justify-content: center;
      }

      .painting {
        width: calc(var(--painting-width) / var(--painting-scalefactor));
        height: calc(var(--painting-height) / var(--painting-scalefactor));
        position: relative;
        transition-property: transform;
        will-change: transform;
      }

      .painting.translate {
        transform: translate(0px, 0px);
      }

      .about {
        position: absolute;
        bottom: -2em;
        padding: 1em;
        font-size: large;
        font-family: monospace;
        font-weight: bold;
      }

      .about a {
        color: #ffffef;
      }

      @media screen and (max-width: 400px) {
        :host {
          padding: 1.5em 0;
          --painting-margin: 1em;
        }

        .paintings {
          --painting-scalefactor: 1.3;
          grid-column-gap: 1.3em;
          grid-row-gap: 1.3em;
        }
      }
    </style>"
`;

exports[`overview-page renders correctly with paintings 1`] = `
"
      <div class="paintings">
        <paint-new-paint-button class="painting translate"></paint-new-paint-button>
        <paint-paint-button class="painting translate" style="transform:translate(0px, 0px);" imageurl="">
              <paint-icon-button role="button" slot="addons" class="delete"><svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 352 512">
  <path d="M242.72 256l100.07-100.07c12.28-12.28 12.28-32.19 0-44.48l-22.24-22.24c-12.28-12.28-32.19-12.28-44.48 0L176 189.28 75.93 89.21c-12.28-12.28-32.19-12.28-44.48 0L9.21 111.45c-12.28 12.28-12.28 32.19 0 44.48L109.28 256 9.21 356.07c-12.28 12.28-12.28 32.19 0 44.48l22.24 22.24c12.28 12.28 32.2 12.28 44.48 0L176 322.72l100.07 100.07c12.28 12.28 32.2 12.28 44.48 0l22.24-22.24c12.28-12.28 12.28-32.19 0-44.48L242.72 256z"></path>
</svg></paint-icon-button>
              
            </paint-paint-button><paint-paint-button class="painting translate" style="transform:translate(0px, 0px);" imageurl="">
              <paint-icon-button role="button" slot="addons" class="delete"><svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 352 512">
  <path d="M242.72 256l100.07-100.07c12.28-12.28 12.28-32.19 0-44.48l-22.24-22.24c-12.28-12.28-32.19-12.28-44.48 0L176 189.28 75.93 89.21c-12.28-12.28-32.19-12.28-44.48 0L9.21 111.45c-12.28 12.28-12.28 32.19 0 44.48L109.28 256 9.21 356.07c-12.28 12.28-12.28 32.19 0 44.48l22.24 22.24c12.28 12.28 32.2 12.28 44.48 0L176 322.72l100.07 100.07c12.28 12.28 32.2 12.28 44.48 0l22.24-22.24c12.28-12.28 12.28-32.19 0-44.48L242.72 256z"></path>
</svg></paint-icon-button>
              
            </paint-paint-button>
      </div>
      <div class="about"><a href="/about">About this app</a></div>
    <style>
  .animated {
    transition-duration: 0.28s;
    transition-timing-function: cubic-bezier(0.4, 0, 0.2, 1);
  }
</style><style>
      :host {
        display: block;
        background-color: #91b5ff;
        padding: 1.5em 0;
        /* prettier-ignore */
        min-width: calc((var(--painting-width) / var(--painting-scalefactor) + var(--painting-margin)) * 2);
        position: relative;
      }

      .paintings {
        display: grid;
        grid-gap: var(--painting-margin);
        grid-template-columns: repeat(
          auto-fit,
          calc(var(--painting-width) / var(--painting-scalefactor))
        );
        justify-content: center;
      }

      .painting {
        width: calc(var(--painting-width) / var(--painting-scalefactor));
        height: calc(var(--painting-height) / var(--painting-scalefactor));
        position: relative;
        transition-property: transform;
        will-change: transform;
      }

      .painting.translate {
        transform: translate(0px, 0px);
      }

      .about {
        position: absolute;
        bottom: -2em;
        padding: 1em;
        font-size: large;
        font-family: monospace;
        font-weight: bold;
      }

      .about a {
        color: #ffffef;
      }

      @media screen and (max-width: 400px) {
        :host {
          padding: 1.5em 0;
          --painting-margin: 1em;
        }

        .paintings {
          --painting-scalefactor: 1.3;
          grid-column-gap: 1.3em;
          grid-row-gap: 1.3em;
        }
      }
    </style>"
`;
